Week2 [STEP 3] Kobe

devKobe24 opened the pull request:
@Gundy93
 건디, 바로 스텝 3에 도전해봤습니다 :)
 이번에는 정말 재미있는 도전이였습니다.

 ## 고민했던 점
 ### 예외처리와 다른 방법으로 풀어보기
 - 사용자가 5회차 이상의 차수. 즉, 6회차 이상을 입력할 때 어떻게 예외처리를 할지 고민해봤습니다.
 - 이전 사용했던 while 문이 아닌 for문과 if 문을 활용해서 코드를 만들어봤습니다 각각의 장단점이 어떤 점이 있을지 궁굼해서 한 번 만들어 봤습니다.

 ---

 ## 해결되지 않은 점
 ### 예외처리
 - 어떤 부분에서 예외처리를 해야할 지 감이 잡히지 않아서 예외처리 없이 코드를 만들었습니다. 함수의 매개변수인  pickTime에 조건문을 달아 예외처리를 했으면 괜찮았을까요?

 ---

 ## 조언을 얻고 싶은 부분
 ### 예외 처리 부분
 - 예외처리는 정말 중요하다고 생각합니다. 건디는 예외처리를 어떠한 방법으로 생각하고 로직을 짜시는지 궁굼합니다.

Commit:
Week2 [STEP 3] Kobe

 @Gundy93
 건디, 바로 스텝 3에 도전해봤습니다 :)
 이번에는 정말 재미있는 도전이였습니다.

 ## 고민했던 점
 ### 예외처리와 다른 방법으로 풀어보기
 - 사용자가 5회차 이상의 차수. 즉, 6회차 이상을 입력할 때 어떻게 예외처리를 할지 고민해봤습니다.
 - 이전 사용했던 while 문이 아닌 for문과 if 문을 활용해서 코드를 만들어봤습니다 각각의 장단점이 어떤 점이 있을지 궁굼해서 한 번 만들어 봤습니다.

 ---

 ## 해결되지 않은 점
 ### 예외처리
 - 어떤 부분에서 예외처리를 해야할 지 감이 잡히지 않아서 예외처리 없이 코드를 만들었습니다.
 함수의 매개변수인  pickTime에 조건문을 달아 예외처리를 했으면 괜찮았을까요?

 ---

 ## 조언을 얻고 싶은 부분
 ### 예외 처리 부분
 - 예외처리는 정말 중요하다고 생각합니다.
 건디는 예외처리를 어떠한 방법으로 생각하고 로직을 짜시는지 궁굼합니다.

Gundy93 commented on CodeStarterCamp_Week2/main.swift:
```diff
@@ -8,5 +8,9 @@
+//matchedLottoNumbers(myLottoNumbers: [1,2,3,4,5,6])
+
+// MARK: - STEP 3
+makeLottoWinnerTimeAndNumbers(pickTime: "5회차")
```
보통 코드를 읽을 때 `main`부터 읽지않을까하는 생각이 듭니다.

처음보는 사람 입장에서 `main`에서 호출되는 함수는 이것 하나뿐인데, 이걸 호출하면 몇번의 로또회차가 뽑히는지는 모르겠지만 일단 로또도 돌리고, 5회차의 번호도 가져오는 것 같습니다. 척 보기에도 하는 일이 많은 함수인 것 같습니다.

적어도 뽑는 것과 회차조회는 서로 분리가 되어있어야할 것 같습니다.

그리고 `pickTime`이라는 이름의 전달인자 레이블은 `Int` 타입의 값을 받을 것 같은데 `String` 타입의 값을 받고 있네요! 오타라도 난다면 문제가 될 것 같습니다. 🤔

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -0,0 +1,43 @@
+	var randomNumbers: Set<Int> = []
+	var times: Int = 0
+	
+	for _ in 0... {
```
현재 범위를 `0...`으로 끝을 명시해주지 않았는데요 와일드카드패턴을 통해 반복되면서 바뀔 시퀀스의 요소는 사용하지 않는데 범위가 정해져 있지 않는 것은 `while true`와 별 차이가 없는 것 같습니다.

물론 `while true`만큼이나 위험한 선택이라는 생각도 듭니다.

PR 메시지에 적으신대로 `for`문의 장단점을 확인하시려면 `for`문에 알맞게 로직을 구현하셔야할 것 같습니다. 정확하게 몇회를 반복할 것인지에 대해 생각해보시는 게 좋겠습니다.

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -0,0 +1,43 @@
+	var times: Int = 0
+	
+	for _ in 0... {
+		let randomNumber = Int.random(in: 1...45)
+		if !randomNumbers.contains(randomNumber) {
+			randomNumbers.insert(randomNumber)
+		}
+		
+		if randomNumbers.count == 6 {
+			times += 1
+			
+			let lottoNumbers = Array(randomNumbers)
```
로또번호를 만드는 로직은 `step2`에서 구현해놓은 함수를 호출해야 될 것 같습니다! 연장선상에 있는 스텝이기 때문입니다!

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -0,0 +1,43 @@
+			
+			let key = String(times) + "회차"
+			
+			lottoWinnerTimeAndNumbers = [key: lottoNumbers]
```
매번 `lottoWinnerTimeAndNumbers` 딕셔너리는 `[key: lottoNumbers]`로 변경되고 있어서 한 회차씩밖에 접근할 수 없습니다.

이 말은 2회차를 뽑으면 1회차 기록이 없어진다는 의미입니다.

게다가 다음과 같이 매번 호출시마다 회차의 값이 바뀌게 되겠습니다

<img width="570" alt="스크린샷 2023-03-19 오후 10 47 24" src="https://user-images.githubusercontent.com/106914201/226179320-c49ecc27-10c3-4a29-856d-774f839b510f.png">

몇 번을 호출하더라도 같은 회차에 대한 조회는 같은 값을 출력해야 맞을 것 같습니다

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -0,0 +1,43 @@
+
+func makeLottoWinnerTimeAndNumbers(pickTime: String) {
+	var randomNumbers: Set<Int> = []
+	var times: Int = 0
```
만약 이 단어를 사용한다면 이 변수는 회차에 대한 값이므로 단수형인 `time`이 더 적절할 것 같습니다.

Gundy93 reviewed (commented):
코비님 이번 스텝을 즐겁게 하셨다니 아주 좋습니다 👍

코딩은 재미가 있어야죠!

> 사용자가 5회차 이상의 차수. 즉, 6회차 이상을 입력할 때 어떻게 예외처리를 할지 고민해봤습니다.

입력한 회차가 적절하지 않다면 안내문구를 출력하는 것도 좋을 것 같습니다

> 이전 사용했던 while 문이 아닌 for문과 if 문을 활용해서 코드를 만들어봤습니다 각각의 장단점이 어떤 점이 있을지 궁굼해서 한 번 만들어 봤습니다.

이 부분은 공식문서의 [Control Flow](https://docs.swift.org/swift-book/documentation/the-swift-programming-language/controlflow) 파트를 읽어보면 좋겠네요

> 어떤 부분에서 예외처리를 해야할 지 감이 잡히지 않아서 예외처리 없이 코드를 만들었습니다. 함수의 매개변수인 pickTime에 조건문을 달아 예외처리를 했으면 괜찮았을까요?

앞선 코멘트를 참고해주세요!

> 예외처리는 정말 중요하다고 생각합니다. 건디는 예외처리를 어떠한 방법으로 생각하고 로직을 짜시는지 궁굼합니다.

저는 보통 `Optional`이나 `Bool`을 고려하는 편입니다. 특정 값에 대한 검증을 하는 함수 등을 만들어 적절하지 못한 값이라면 `nil`을 반환한다거나 `false`를 반환하는 방식으로 말이죠

이번 스텝 수정도 화이팅입니다!

Commit:
Week2 [STEP 3] Kobe

 @Gundy93
 건디, 정말 많은 시행착오 후에 성공했습니다 ㅠㅠ
 수정 후 PR 보내드립니다.

 ## 고민했던 점
 ### 전역 변수와 지역 변수의 차이점
 - 전역 변수와 지역 변수의 차이점에 따라 값이 어떻게 바뀌고 쓰이는지에 대하여 많은 고민과 공부를 하게 되었습니다.

 ---

 ## 해결되지 않은 점
 ### 이번에는 예외처리를 switch 문의 default로 해서 처리해봤습니다.
 - 이게 맞는 예외처리인지 또 다른 예외처리에는 무엇이 있는지 궁굼합니다.

 ---

 ## 조언을 얻고 싶은 부분
 ### 걱정이 많이 되는 부분이 있습니다.
 - 건디 저는 이제 스텝 2를 하고 있습니다. 지난번 건디께서 스텝 3만 해도 비빌만 하다고 하셨는데 ㅠㅠ
 저는 아직 스텝 3 시작도 못했는데 커리어 캠프에 떨어질까 너무 너무 걱정이 많습니다.
 회사는 4월 6일에 그만두기로 했고 이제 뒤가 없는데 너무 떨립니다 ㅠㅠ

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -0,0 +1,59 @@
+
+import Foundation
+
+func makeNumbers() -> [Int] {
+	var randomNumbers: Set<Int> = []
+	
+	while randomNumbers.count != 6 {
+		let randomNumber = Int.random(in: 1...45)
+		randomNumbers.insert(randomNumber)
+	}
+	
+	let lottoNumbers = Array(randomNumbers)
+
+	return lottoNumbers
+}
```
step2에서 정확히 같은 함수를 만들었는데요, 또 만든 이유가 있을까요?

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -0,0 +1,59 @@
+	return lottoNumbers
+}
+
+let first = makeNumbers()
+let second = makeNumbers()
+let third = makeNumbers()
+let fourth = makeNumbers()
+let fifth = makeNumbers()
```
다섯개의 전역변수를 만들어주셨는데, 그렇다면 6회차, 7회차가 생길 때마다 새 전역변수가 추가되는 걸까요?

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -0,0 +1,59 @@
+let fifth = makeNumbers()
+
+func confirmLottoNumbers(time: Int) {
+	let dictionary: [Int: [Int]] = [
+		1: first,
+		2: second,
+		3: third,
+		4: fourth,
+		5: fifth
+	]
```
이 딕셔너리가 함수 내부의 지역변수가 아닌, 전역변수여야할 것 같습니다.

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -0,0 +1,59 @@
+		5: fifth
+	]
+	
+	guard let firstTime = dictionary[1] else { return }
+	guard let secondTime = dictionary[2] else { return }
+	guard let thirdTime = dictionary[3] else { return }
+	guard let fourthTime = dictionary[4] else { return }
+	guard let fifthTime = dictionary[5] else { return }
```
1부터 5까지 모든 key에 대해 옵셔널 바인딩을 하고 있는데, 매개변수로 전달된 `time`만 조회해보면 되지 않을까 싶습니다

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -0,0 +1,59 @@
+	guard let fourthTime = dictionary[4] else { return }
+	guard let fifthTime = dictionary[5] else { return }
+	
+	let firstTimeWinnerNumbers = firstTime.map { String($0) }.joined(separator: ", ")
+	let secondTimeWinnerNumbers = secondTime.map { String($0) }.joined(separator: ", ")
+	let thirdTimeWinnerNumbers = thirdTime.map { String($0) }.joined(separator: ", ")
+	let fourthTimeWinnerNumbers = fourthTime.map { String($0) }.joined(separator: ", ")
+	let fifthTimeWinnerNumbers = fifthTime.map { String($0) }.joined(separator: ", ")
```
이것도 앞선 코멘트와 같은 맥락입니다!

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -0,0 +1,59 @@
+	let fifthTimeWinnerNumbers = fifthTime.map { String($0) }.joined(separator: ", ")
+	
+	switch time {
+	case 1: print("\(time)회차 로또 당첨 번호는 \(firstTimeWinnerNumbers)입니다.")
+	case 2: print("\(time)회차 로또 당첨 번호는 \(secondTimeWinnerNumbers)입니다.")
+	case 3: print("\(time)회차 로또 당첨 번호는 \(thirdTimeWinnerNumbers)입니다.")
+	case 4: print("\(time)회차 로또 당첨 번호는 \(fourthTimeWinnerNumbers)입니다.")
+	case 5: print("\(time)회차 로또 당첨 번호는 \(fifthTimeWinnerNumbers)입니다.")
```
코드가 상당히 반복사용되고 있는데! 이 중복을 줄일 수 있을 것 같아요

Gundy93 reviewed (commented):
제 생각에는 지금 하나의 함수로 너무 많은 일을 해결하려고 하시는 것 같습니다.

1. 전역변수인 딕셔너리
2. step2에서 구현한 로또번호 생성함수를 사용해 새 회차에 로또번호를 저장하는 함수
3. 매개변수로 회차를 받아 저장된 로또번호를 확인하는 함수

이런 식으로 구분해서 구현한다면 어떨까요?

`main.swift`에서 2번함수를 `for-in` 반복문을 사용해 5회 호출한 다음

3번 함수를 사용해서 특정 회차의 번호를 조회하면 어떨까요?

지금 이야기 한 것은 최소한의 구분이고, 더 세분화해보는 것도 좋습니다!

Commit:
Week2 [STEP 3] Kobe

 @Gundy93
 건디, 이게 마지막 PR이 될 것 같네요 ㅠㅠ
 5주라는 시간이 이렇게 짧을 줄 몰랐어요 ㅠㅠ
 아마 건디와 함께 한 코드 리뷰가 너무 재미있어서 짧게 느껴졌나봅니다.
 너무 즐겁고 행복한 코딩을 알려주셔서 감사합니다 건디.
 피드백 반영하여 수정하여 마지막 PR 보내드립니다.

 ## 고민했던 점
 ### 무작위의 숫자가 변하지 않고 딕셔너리에 남아 있게 하는 것
 - 전역 변수, 지역 변수의 목적성에 따라 사용해야 하는 것을 많이 고민하고 피드백에 따라 딕셔너리를 전역변수로 만들어보았습니다.

 ---

 ## 해결되지 않은 점
 ### 해결되지 않은 점이 이번 스텝에는 없습니다.

 ---

 ## 조언을 얻고 싶은 부분
 ### 건디와 더 많은 코드리뷰를 하고싶은데 방법이 없을까요?? ㅠㅠ
 - 건디와 더 많은 소통과 코드리뷰를 하고싶은데 방법이 없을까요??
 정말 너무 좋은 학습도우미인 건디를 만나 하루 하루 성장하는 저를 보면서 너무 즐겁게 코딩했습니다.
 건디와 더 오랜시간 코드리뷰를 하고싶은데 방법이 없을까요?? ㅠㅠ

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -7,53 +7,71 @@
-	case 5: print("\(time)회차 로또 당첨 번호는 \(fifthTimeWinnerNumbers)입니다.")
-	default: print("1회차 이하의 당첨 번호 혹은 5회차 이상의 당첨 번호는 없습니다.")
+// 2. Step2에서 구현한 로또번호 생성함수를 사용해 새 회차에 로또번호를 저장하는 함수
+func matchTimeAndLottoNumbers(time: Int) {
+	let lottoNumbers = makeLottoNumbers()
+	winnerTimeAndLottoNumbers[time] = lottoNumbers
+}
```
현재는 특정 회차를 매개변수로 전달받아서 저장하게 되는데요!

이 방법 대신 기존의 마지막 회차의 다음 회차로 저장하는 로직으로 구현해보면 어떨까요?

Gundy93 commented on CodeStarterCamp_Week2/STEP3.swift:
```diff
@@ -7,53 +7,71 @@
+}
+// 3. 매개변수로 회차를 받아 저장된 로또번호를 확인하는 함수
+func confirmWinnerLottoNumbers(time: Int) {
+	if time <= 5 {
```
현재는 5회차까지만 고정적으로 뽑기 때문에 `time`의 조건이 `5`이하여야 제대로 조회할 수 있지만, 그 이후로도 계속 추첨을 진행해 6,7,...회차까지 진행된다면 이 조건으로 감당할 수 없을 거에요!

딕셔너리에 저장된 마지막 회차, 혹은 회차의 개수 등을 확인해보는 방법이 필요할 것 같습니다

Gundy93 reviewed (commented):
거의 다 수정이 된 것 같아요! 바라던 모습과 근접합니다 👍

기능적인 부분은 고칠 게 없지만, 현재는 딱 지금의 조건인 5회를 뽑고 조회하는 기능으로만 구현이 되어있습니다

조금 더 범용성이 좋게 살짝만 손보고 마무리하면 좋을 것 같아요

코비님 5주간 수고하셨습니다 😃

> 건디와 더 많은 코드리뷰를 하고싶은데 방법이 없을까요?? ㅠㅠ

커리어 스타터 캠프를 신청하시면 됩니다!

커리어 스타터 캠프의 리뷰는 현업자가 주로 해주시지만 캠프 초반에는 학습도우미가 리뷰하게 됩니다.

따라서 커리어 스타터 캠프에 당당히 합격해서 만나요 코비님!

마지막 수정까지 화이팅입니다 💪
